From 30fa13721d34a6755b17e87b19d32a8b3c75a535 Mon Sep 17 00:00:00 2001
From: Frank Meisenbach <fmeisenbach@hilscher.com>
Date: Thu, 18 Mar 2021 23:49:35 +0100
Subject: [PATCH 1/2] atheros.c: Fix the vddio-1p8v and smartEEE quirk of
 Atheros ar803x Ethernet-Phy.

Note: These fixes are based on the Linux driver.
---
 drivers/net/phy/atheros.c | 119 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 119 insertions(+)

diff --git a/drivers/net/phy/atheros.c b/drivers/net/phy/atheros.c
index 3783d155e7..a93a31c87d 100644
--- a/drivers/net/phy/atheros.c
+++ b/drivers/net/phy/atheros.c
@@ -7,6 +7,7 @@
  */
 #include <common.h>
 #include <phy.h>
+#include <malloc.h>
 
 #define AR803x_PHY_DEBUG_ADDR_REG	0x1d
 #define AR803x_PHY_DEBUG_DATA_REG	0x1e
@@ -17,6 +18,12 @@
 #define AR803x_DEBUG_REG_0		0x0
 #define AR803x_RGMII_RX_CLK_DLY		0x8000
 
+struct at803x_priv {
+	u32 quirks;
+};
+#define AT803X_QUIRKS_SMART_EEE_DISABLE	(1 << 1)
+#define AT803X_QUIRKS_VDDIO_1P8V		(1 << 2)
+
 static int ar8021_config(struct phy_device *phydev)
 {
 	phy_write(phydev, MDIO_DEVAD_NONE, 0x00, 0x1200);
@@ -27,8 +34,94 @@ static int ar8021_config(struct phy_device *phydev)
 	return 0;
 }
 
+static int at803x_debug_reg_read(struct phy_device *phydev, u16 reg)
+{
+	int ret;
+
+	ret = phy_write(phydev, MDIO_DEVAD_NONE, AR803x_PHY_DEBUG_ADDR_REG, reg);
+	if (ret < 0)
+		return ret;
+
+	return phy_read(phydev, MDIO_DEVAD_NONE, AR803x_PHY_DEBUG_DATA_REG);
+}
+
+static int at803x_debug_reg_mask(struct phy_device *phydev, u16 reg, u16 clear, u16 set)
+{
+	u16 val;
+	int ret;
+
+	ret = at803x_debug_reg_read(phydev, reg);
+	if (ret < 0)
+		return ret;
+
+	val = ret & 0xffff;
+	val &= ~clear;
+	val |= set;
+
+	return phy_write(phydev, MDIO_DEVAD_NONE, AR803x_PHY_DEBUG_DATA_REG, val);
+}
+
+#define AT803X_DEBUG_REG_31		0x1f
+#define AT803X_VDDIO_1P8V_EN	(1 << 3)
+static inline int at803x_vddio_1p8v(struct phy_device *phydev, int enable)
+{
+	int ret;
+
+	if (enable)
+		ret = at803x_debug_reg_mask(phydev, AT803X_DEBUG_REG_31, 0, AT803X_VDDIO_1P8V_EN);
+	else
+		ret = at803x_debug_reg_mask(phydev, AT803X_DEBUG_REG_31, AT803X_VDDIO_1P8V_EN, 0);
+
+	return ret;
+}
+
+#define AT803X_SMART_EEE_CTRL3	0x805D
+#define AT803X_LPI_EN			(1 << 8)
+static int at803x_smart_eee(struct phy_device *phydev, int enable)
+{
+	int value;
+	int ret;
+
+	/* 5.1.11 Smart_eee control3 */
+	value = phy_read_mmd(phydev, MDIO_MMD_PCS, AT803X_SMART_EEE_CTRL3);
+
+	(enable) ? (value |= AT803X_LPI_EN) : (value &= ~AT803X_LPI_EN);
+
+	ret = phy_write_mmd(phydev, MDIO_MMD_PCS, AT803X_SMART_EEE_CTRL3, value);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+static int ar803x_quirks(struct phy_device *phydev)
+{
+	struct at803x_priv *priv = phydev->priv;
+	struct ofnode_phandle_args phandle_args;
+
+	if (dev_read_phandle_with_args(phydev->dev, "phy-handle", NULL, 0, 0, &phandle_args)) {
+		debug("Failed to find phy-handle");
+		return -ENODEV;
+	}
+
+	if (ofnode_read_bool(phandle_args.node, "at803x,eee-disabled"))
+		priv->quirks |= AT803X_QUIRKS_SMART_EEE_DISABLE;
+
+	if (ofnode_read_bool(phandle_args.node, "at803x,vddio-1p8v"))
+		priv->quirks |= AT803X_QUIRKS_VDDIO_1P8V;
+
+	return 0;
+}
+
 static int ar8031_config(struct phy_device *phydev)
 {
+	struct at803x_priv *priv = phydev->priv;
+	int ret;
+
+	ret = ar803x_quirks(phydev);
+	if (ret < 0)
+		return ret;
+
 	if (phydev->interface == PHY_INTERFACE_MODE_RGMII_TXID ||
 	    phydev->interface == PHY_INTERFACE_MODE_RGMII_ID) {
 		phy_write(phydev, MDIO_DEVAD_NONE, AR803x_PHY_DEBUG_ADDR_REG,
@@ -45,6 +138,18 @@ static int ar8031_config(struct phy_device *phydev)
 			  AR803x_RGMII_RX_CLK_DLY);
 	}
 
+	if (priv->quirks & AT803X_QUIRKS_VDDIO_1P8V) {
+		ret = at803x_vddio_1p8v(phydev, 1);
+		if (ret < 0)
+			return ret;
+	}
+
+	if (priv->quirks & AT803X_QUIRKS_SMART_EEE_DISABLE) {
+		ret = at803x_smart_eee(phydev, 0);
+		if (ret < 0)
+			return ret;
+	}
+
 	phydev->supported = phydev->drv->features;
 
 	genphy_config_aneg(phydev);
@@ -91,6 +196,21 @@ static int ar8035_config(struct phy_device *phydev)
 	return 0;
 }
 
+static int ar803x_probe(struct phy_device *phydev)
+{
+	struct at803x_priv *priv;
+
+	priv = malloc(sizeof(*priv));
+	if (!priv)
+		return -ENOMEM;
+
+	memset(priv, 0, sizeof(*priv));
+
+	phydev->priv = priv;
+
+	return 0;
+}
+
 static struct phy_driver AR8021_driver =  {
 	.name = "AR8021",
 	.uid = 0x4dd040,
@@ -106,6 +226,7 @@ static struct phy_driver AR8031_driver =  {
 	.uid = 0x4dd074,
 	.mask = 0xffffffef,
 	.features = PHY_GBIT_FEATURES,
+	.probe = ar803x_probe,
 	.config = ar8031_config,
 	.startup = genphy_startup,
 	.shutdown = genphy_shutdown,
-- 
2.17.1

